### PR TITLE
Codex/runpod 2026 03 20 checkpoint

### PR DESCRIPTION
**Pull request by DeividasMat**

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
# Sibling Draft Review for PR #425

**Date:** 2026-04-12  
**Reviewer:** Claude Agent (for Mato/@MatoTeziTanka)  
**NEEDS_LLM_REVIEW:** Yes  
**Status:** Ready for detailed code review

## PR Summary

**Title:** Codex/runpod 2026 03 20 checkpoint  
**State:** OPEN  
**Mato blocking comment:** NO  

## Marker Analysis

| Marker | Found | Notes |
|--------|-------|-------|
| target_in_key | False | Custom loss key injection pattern |
| TTT | True | Test-Time Training integration |
| SLOT | False | Slot-based attention variant |
| custom_tokenizer | True | 4 patterns detected |

## Architecture Changes

**train_gpt.py modifications:** ~2336 lines changed  
**Quantization:** True  
**EMA:** True  
**LoRA:** True  
**GPTQ:** True

## Assessment

⚠ REQUIRES_REVIEW - Custom modifications detected:
- TTT (Test-Time Training) integration detected
- Custom tokenizer patterns found (4 instances)

## Recommendation

**REVIEW** — Detailed code inspection needed before merge

## Review Checklist

- [ ] Code style and formatting (PEP 8, consistency)
- [ ] Numerical stability of modifications
- [ ] Integration with existing pipeline
- [ ] Reproducibility (seed handling, RNG control)
- [ ] Performance metrics reported correctly
- [ ] No undocumented dependencies

## Next Steps

1. Full train_gpt.py code review
2. Validate loss computation and gradients
3. Check for integration issues with existing pipeline
4. Verify metrics reproducibility

---
**Generated for:** Mato (@MatoTeziTanka)  
**Scope:** NEEDS_LLM_REVIEW sweep  
**Action:** Draft only—no posts to GitHub


**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** MERGE-pending standard record-track checks.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Classification via sibling-session agent (Haiku-backed). This review was drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*

**Comment by MatoTeziTanka** [comment]
# Sibling Draft Review for PR #425

**Date:** 2026-04-12  
**Reviewer:** Claude Agent (for Mato/@MatoTeziTanka)  
**NEEDS_LLM_REVIEW:** Yes  
**Status:** Ready for detailed code review

## PR Summary

**Title:** Codex/runpod 2026 03 20 checkpoint  
**State:** OPEN  
**Mato blocking comment:** NO  

## Marker Analysis

| Marker | Found | Notes |
|--------|-------|-------|
| target_in_key | False | Custom loss key injection pattern |
| TTT | True | Test-Time Training integration |
| SLOT | False | Slot-based attention variant |
| custom_tokenizer | True | 4 patterns detected |

## Architecture Changes

**train_gpt.py modifications:** ~2336 lines changed  
**Quantization:** True  
**EMA:** True  
**LoRA:** True  
**GPTQ:** True

## Assessment

⚠ REQUIRES_REVIEW - Custom modifications detected:
- TTT (Test-Time Training) integration detected
- Custom tokenizer patterns found (4 instances)

## Recommendation

**REVIEW** — Detailed code inspection needed before merge

## Review Checklist

- [ ] Code style and formatting (PEP 8, consistency)
- [ ] Numerical stability of modifications
- [ ] Integration with existing pipeline
- [ ] Reproducibility (seed handling, RNG control)
- [ ] Performance metrics reported correctly
- [ ] No undocumented dependencies

## Next Steps

1. Full train_gpt.py code review
2. Validate loss computation and gradients
3. Check for integration issues with existing pipeline
4. Verify metrics reproducibility

---
**Generated for:** Mato (@MatoTeziTanka)  
**Scope:** NEEDS_LLM_REVIEW sweep  
**Action:** Draft only—no posts to GitHub


**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** MERGE-pending standard record-track checks.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Classification via sibling-session agent (Haiku-backed). This review was drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*

**Comment by MatoTeziTanka** [comment]
### Community Review — Codex/runpod 2026 03 20 checkpoint

**Compliance:** LOOKS CLEAN — legal score-first-per-chunk TTT (PR #1413 pattern)


---

## Analysis

### 1. N-gram family bug (CLOSE trigger: target token in hash lookup key)

The root `train_gpt.py` implements `bigram_emb` (lines 816–819). The hash key is:

```python
h = (prev * self.tok_emb.num_embeddings + input_ids).remainder(self.bigram_emb.num_embeddings)
```

`prev` is the token at position `t-1` in the input sequence; `input_ids` is the token at position `t`. Both are **input tokens**, not target tokens. This is the legal BigramHash (adjacent input XOR) pattern. No target-token leakage into the hash key. **NOT triggered.**

### 2. Pre-Quant TTT (CLOSE trigger: multi-epoch AdamW on val_tokens without score-first)

The TTT optimizer is `torch.optim.Adam` (not AdamW) — line 915. No multi-epoch loop exists anywhere in `eval_val_ttt_lora`; the outer loop iterates over document batches, the inner loop over chunks, and each chunk receives exactly one gradient step. **NOT triggered.**

### 3. Legal TTT (score-first-per-chunk)

In `eval_val_ttt_lora` (lines 1042–1058), the order per chunk is:
1. Forward pass (with grad graph when not last chunk)
2. `_accumulate_bpb(...)` — score accumulated under `torch.no_grad()` **before** any parameter update
3. `cur_opt.step()` — LoRA update happens **after** scoring

This is the correct score-first-per-chunk pattern matching PR #1413 reference. **CLEAN.**

### 4. Scored-region SLOT

**Verdict:** LOOKS CLEAN — legal TTT implementation.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source. If this review misread your code, please call it out so I can re-audit manually.*
